sympy/core/basic.py

```diff
@@ -1228,31 +1229,72 @@ def has(self, *patterns):
         False
 
         """
-        return any(self._has(pattern) for pattern in patterns)
+        return self._has(iterargs, *patterns)
 
-    def _has(self, pattern):
-        """Helper for .has()"""
-        from .function import UndefinedFunction, Function
-        if isinstance(pattern, UndefinedFunction):
-            return any(pattern in (f, f.func)
-                       for f in self.atoms(Function, UndefinedFunction))
+    @cacheit
+    def has_free(self, *patterns):
+        """return True if self has object(s) ``x`` as a free expression
+        else False.
 
-        if isinstance(pattern, BasicMeta):
-            subtrees = _preorder_traversal(self)
-            return any(isinstance(arg, pattern) for arg in subtrees)
+        Examples
+        ========
 
-        pattern = _sympify(pattern)
+        >>> from sympy import Integral, Function
+        >>> from sympy.abc import x, y
+        >>> f = Function('f')
+        >>> g = Function('g')
+        >>> expr = Integral(f(x), (f(x), 1, g(y)))
+        >>> expr.free_symbols
+        {y}
+        >>> expr.has_free(g(y))
+        True
+        >>> expr.has_free(*(x, f(x)))
+        False
 
-        _has_matcher = getattr(pattern, '_has_matcher', None)
-        if _has_matcher is not None:
-            match = _has_matcher()
-            return any(match(arg) for arg in _preorder_traversal(self))
-        else:
-            return any(arg == pattern for arg in _preorder_traversal(self))
+        This works for subexpressions and types, too:
+
+        >>> expr.has_free(g)
+        True
+        >>> (x + y + 1).has_free(y + 1)
+        True
+
+        """
+        return self._has(iterfreeargs, *patterns)
+
+    def _has(self, iterargs, *patterns):
+        # separate out types and unhashable objects
+        type_set = set()  # only types
+        p_set = set()  # hashable non-types
+        for p in patterns:
+            if isinstance(p, BasicMeta):
+                type_set.add(p)
+                continue
+            if not isinstance(p, Basic):
+                try:
+                    p = _sympify(p)
+                except SympifyError:
+                    continue  # Basic won't have this in it
+            p_set.add(p)  # fails if object defines __eq__ but
+                          # doesn't define __hash__
+                                  #
+        for i in iterargs(self):  #
+            if i in p_set:        # <--- here, too
+                return True
+            if type(i) in type_set or any(isinstance(i, t) for t in type_set):
+                return True
```

Comment:
When there are lots of types this can be slow:
```python
In [1]: p = 1 + x**2

In [2]: for _ in range(10):
   ...:     p = p**2 + 1/p
   ...: 

In [3]: types = cos, sin, tan, cosh, sinh, tanh, exp, log, cot, sec, csc, coth, sinh, csch

In [4]: %time p.has(*types)
CPU times: user 76 ms, sys: 0 ns, total: 76 ms
Wall time: 74.8 ms
Out[4]: False
```
It's a bit faster to use a tuple:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index acf28a8..20ac83c 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -1276,11 +1276,12 @@ def _has(self, iterargs, *patterns):
                     continue  # Basic won't have this in it
             p_set.add(p)  # fails if object defines __eq__ but
                           # doesn't define __hash__
+        types = tuple(type_set)
                                   #
         for i in iterargs(self):  #
             if i in p_set:        # <--- here, too
                 return True
-            if type(i) in type_set or any(isinstance(i, t) for t in type_set):
+            if isinstance(i, types):
                 return True
 
         # use matcher if defined, e.g. operations defines
```
That gives:
```python
In [4]: %time p.has(*types)
CPU times: user 52 ms, sys: 0 ns, total: 52 ms
Wall time: 52.8 ms
Out[4]: False
```